monai/data/wsi_reader.py

```diff
@@ -0,0 +1,310 @@
+            location: (x_min, y_min) tuple giving the top left pixel in the level 0 reference frame,
+            or list of tuples (default=(0, 0))
+            size: (height, width) tuple giving the patch size, or list of tuples (default to full image size)
+            This is the size of image at the given level (`level`)
+            level: the level number, or list of level numbers (default=0)
+            dtype: the data type of output image
+            mode: the output image mode 'RGB' or 'RGBA'
+
+        Returns:
+            a tuples, where the first element is an image [CxHxW], and second element is a dictionary of metadata
+        """
+        # Verify magnification level
+        if level is None:
+            level = self.level
+        max_level = self._get_level_count(wsi) - 1
+        if level > max_level:
+            raise ValueError(f"The maximum level of this image is {max_level} while level={level} is requested)!")
+
+        # Verify location
+        if location is None:
+            location = (0, 0)
+        wsi_size = self._get_size(wsi, level)
+        if location[0] > wsi_size[0] or location[1] > wsi_size[1]:
+            raise ValueError(f"Location is outside of the image: location={location}, image size={wsi_size}")
+
+        # Verify size
+        if size is None:
+            if location != (0, 0):
+                raise ValueError("Patch size should be defined to exctract patches.")
+            size = self._get_size(wsi, level)
+        else:
+            if size[0] <= 0 or size[1] <= 0:
+                raise ValueError(f"Patch size should be greater than zero, provided: patch size = {size}")
+
+        # Extract a patch or the entire image
+        patch = self._get_patch(wsi, location=location, size=size, level=level, dtype=dtype, mode=mode)
+
+        # Verify patch image
+        patch = self._verify_output(patch, mode)
+
+        # Set patch-related metadata
+        metadata = self._get_metadata(wsi=wsi, patch=patch, location=location, size=size, level=level)
+
+        return patch, metadata
+
+    def verify_suffix(self, filename: Union[Sequence[PathLike], PathLike]) -> bool:
+        """
+        Verify whether the specified file or files format is supported by WSI reader.
+
+        Args:
+            filename: file name or a list of file names to read.
+                if a list of files, verify all the suffixes.
+        """
+        return is_supported_format(filename, self.supported_formats)
+
+    def _verify_output(self, patch: np.ndarray, mode: str):
+        """
+        Verify output image patch
+        """
+        # check if the image has three dimensions (2D + color)
+        if patch.ndim != 3:
+            raise ValueError(
+                f"The image dimension should be 3 but has {patch.ndim}. "
+                "`WSIReader` is designed to work only with 2D images with color channel."
+            )
+
+        return patch
+
+
+class WSIReader(BaseWSIReader):
```

Comment:
I feel we can remove or deprecate this `WSIReader` class directly, no need to maintain it anymore if we define `CuCIMWSIReader` and other subclass readers.
@wyli @ericspod what do you think?

Thanks in advance.

Reply:
I believe it's good to keep it here. It provides not only some backward compatibilities but also a simpler interface for people who doesn't know about underlying libraries like cucim, openslide, etc.

Reply:
OK, I see your point. Let's keep it for now.

Thanks.